src/passes/PostAssemblyScript.cpp

```diff
@@ -0,0 +1,530 @@
+    // throw(..., local.get(X, ...), ...) ?
+    for (auto* operand : curr->operands) {
+      if (auto* localGet = operand->dynCast<LocalGet>()) {
+        escapes.insert(localGet);
+        break;
+      }
+    }
+  }
+
+  void eliminateRetain(Expression** location) {
+    // assumes local.set(X, __retain(...))
+    auto* localSet = (*location)->cast<LocalSet>();
+    auto* call = localSet->value->cast<Call>();
+    assert(call->target == RETAIN);
+    assert(call->operands.size() == 1);
+    localSet->value = call->operands[0];
+  }
+
+  void eliminateRelease(Expression** location) {
+    // assumes __release(local.get(X, ...))
+    if (auto* call = (*location)->dynCast<Call>()) {
+      assert(call->target == RELEASE);
+      assert(call->operands.size() == 1);
+      assert(call->operands[0]->is<LocalGet>());
+      Builder builder(*getModule());
+      *location = builder.makeNop();
+    } else {
+      assert((*location)->is<Nop>()); // already replaced
+    }
+  }
+
+  // Tests if a retain reaches an escape and thus is
+  // considered necessary.
+  bool testReachesEscape(LocalSet* retain, AliasGraph& graph) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      if (releases.find(localGet) != releases.end()) {
+        continue;
+      }
+      if (escapes.find(localGet) != escapes.end()) {
+        return true;
+      }
+      for (auto* localSet : graph.getInfluences[localGet]) {
+        if (testReachesEscape(localSet, graph)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  // Collects all reachable releases of a retain
+  void collectReleases(LocalSet* retain,
+                       AliasGraph& graph,
+                       std::unordered_set<Expression**>& found) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      auto foundRelease = releases.find(localGet);
+      if (foundRelease != releases.end()) {
+        found.insert(foundRelease->second);
+      } else {
+        for (auto* localSet : graph.getInfluences[localGet]) {
+          collectReleases(localSet, graph, found);
+        }
+      }
+    }
+  }
+
+  // Given a retain, gets the retained expression
+  static Expression* getRetainedExpression(LocalSet* retain) {
+    // assumes local.set(X, __retain(...))
+    auto* retainCall = retain->value->cast<Call>();
+    assert(retainCall->operands.size() == 1);
+    return retainCall->operands[0];
+  }
+
+  // Tests if a retained value originates at an allocation
+  // and thus is considered necessary.
+  bool testRetainsAllocation(Expression* retained, AliasGraph& graph) {
+    if (auto* call = retained->dynCast<Call>()) {
+      if (call->target == ALLOC || call->target == ALLOCARRAY) {
+        return true;
+      }
+    } else {
+      if (auto* localGet = retained->dynCast<LocalGet>()) {
+        for (auto* localSet : graph.getSetses[localGet]) {
+          if (localSet != nullptr &&
+              testRetainsAllocation(localSet->value, graph)) {
```

Comment:
ditto recursion risk. here i guess if a limit is reached, can assume it does retain?